src/suites/cts/textures/zero_init.spec.ts

```diff
@@ -0,0 +1,1244 @@
+
+    const { mipRange, sliceRange } = subresourceRange;
+
+    const largestWidth = this.textureWidth >> mipRange.base;
+    const largestHeight = this.textureHeight >> mipRange.base;
+
+    const bytesPerPixel = kTextureFormatInfo[this.params.format].bytes;
+    const minBytesPerRow = largestWidth * bytesPerPixel;
+    const bytesPerRow = Math.ceil(minBytesPerRow / 256) * 256;
+
+    const [buffer, mapping] = this.device.createBufferMapped({
+      size: bytesPerRow * largestHeight,
+      usage: C.BufferUsage.CopySrc,
+    });
+
+    const pixelData = new Uint8Array(this.gpuDataRepresentation[state]);
+    assert(pixelData.byteLength === bytesPerPixel);
+    for (let row = 0; row < largestHeight; ++row) {
+      for (let byteOffset = 0; byteOffset < minBytesPerRow; byteOffset += bytesPerPixel) {
+        const pixel = new Uint8Array(mapping, row * bytesPerRow + byteOffset, bytesPerPixel);
+        pixel.set(pixelData);
+      }
+    }
+    buffer.unmap();
+
+    const commandEncoder = this.device.createCommandEncoder();
+    for (let level = mipRange.base; level < mipRange.base + mipRange.count; ++level) {
+      for (let layer = sliceRange.base; layer < sliceRange.base + sliceRange.count; ++layer) {
+        const width = this.textureWidth >> level;
+        const height = this.textureHeight >> level;
+
+        commandEncoder.copyBufferToTexture(
+          {
+            buffer, bytesPerRow, rowsPerImage: height,
+            // @ts-ignore
+            rowPitch: bytesPerRow, imageHeight: height
+          },
+          { texture, mipLevel: level, arrayLayer: layer },
+          { width, height, depth: 1 }
+        )
+      }
+    }
+    this.device.defaultQueue.submit([commandEncoder.finish()]);
+    buffer.destroy();
+
+    // Make sure it actually initialized the subresources.
+    this.checkContents(texture, state, subresourceRange);
+  }
+
+  discardTexture(texture: GPUTexture,
+                 subresourceRange: SubresourceRange) {
+    const commandEncoder = this.device.createCommandEncoder();
+
+    for (const desc of this.generateTextureViewDescriptorsForRendering(this.params.aspect, subresourceRange)) {
+      if (kTextureFormatInfo[this.params.format].color) {
+        commandEncoder.beginRenderPass({
+          colorAttachments: [{
+            attachment: texture.createView(desc),
+            storeOp: C.StoreOp.Clear,
+            loadValue: C.LoadOp.Load,
+          }],
+        }).endPass();
+      } else {
+        commandEncoder.beginRenderPass({
+          colorAttachments: [],
+          depthStencilAttachment: {
+            attachment: texture.createView(desc),
+            depthStoreOp: C.StoreOp.Clear,
+            depthLoadValue: C.LoadOp.Load,
+            stencilStoreOp: C.StoreOp.Clear,
+            stencilLoadValue: C.LoadOp.Load,
+          }
+        }).endPass();
+      }
+    }
+    this.device.defaultQueue.submit([commandEncoder.finish()]);
+  }
+}
+
+function* generateParams(): Generator<TestParams> {
```

Comment:
didn't review this in depth but it looks very reasonable. It got me thinking about whether we could possibly express this in a more functional way (more in the style of combine and friends). How about:

```js
new Params()
  .combine(
    options('format', kTextureFormats), //
    options('aspect', kTextureAspects))
  .unless(p => p.aspect === 'depth-only' && !kTextureFormatInfo[p.format].depth)
  .unless(p => p.aspect === 'stencil-only' && !kTextureFormatInfo[p.format].stencil)
  .combine(
    kCreationSizes, //
    options('mipLevelCount', kMipLevelCounts),
    options('sampleCount', kSampleCounts))
  .unless(p => p.sampleCount > 1 && p.mipLevelCount > 1)
  .combine(
    options('uninitializeMethod', kUninitializeMethods)), //
    options('readMethod', kReadMethods))
  .unless(p => {
    const usage = getRequiredTextureUsage(sampleCount, uninitializeMethod, readMethod);
    ...
  })
  ...
```

WDYT? Though I know you tried to make nice types for something similar and it didn't seem to work out :(

Also probably we should have helpers for things like `formatSupportsAspect` and `formatSupportsUsage`.

Reply:
Would be quite nice, though even for this test, that method could get quite complicated since there's so many different combinations that might not be supported.